MINIFICPP-1923 Refactor PutUDP to use asio

martinzink opened the pull request:
MINIFICPP-1939 Enable dual stack listening on ListenTCP and ListenSyslog

---

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

martinzink commented on extensions/standard-processors/tests/unit/PutUDPTests.cpp:
```diff
@@ -24,91 +24,86 @@
+}
 }  // namespace
 
-// Testing the failure relationship is not required, because since UDP in general without guarantees, flow files are always routed to success, unless there is
-// some weird IO error with the content repo.
```
This was not strictly true, because the hostname resolution could fail (added a test-case for that).

fgerlits commented on extensions/standard-processors/processors/PutUDP.cpp:
```diff
@@ -107,48 +98,48 @@ void PutUDP::onTrigger(core::ProcessContext* context, core::ProcessSession* cons
-        logger_->log_error("%s", ec.message());
-        session->transfer(flow_file, Failure);
-      });
+  resolve_hostname()
+      | utils::map(debug_log_resolved_endpoint)
+      | utils::flatMap(send_data_to_endpoint)
+      | utils::map(transfer_to_success)
+      | utils::orElse(transfer_to_failure);
```
👍 it's much more readable this way

fgerlits commented on extensions/standard-processors/processors/PutUDP.cpp:
```diff
@@ -107,48 +98,48 @@ void PutUDP::onTrigger(core::ProcessContext* context, core::ProcessSession* cons
+
+  const auto debug_log_resolved_endpoint = [&hostname, &logger = this->logger_](const udp::resolver::results_type& resolved_query) -> udp::endpoint {
+    if (logger->should_log(core::logging::LOG_LEVEL::debug))
+      core::logging::LOG_WARN(logger) << "resolved " << hostname << " to: " << resolved_query->endpoint();
```
should this be `LOG_DEBUG`?

martinzink replied:
you are right, fixed it https://github.com/apache/nifi-minifi-cpp/pull/1412/commits/45948030831865d2a71d6a47135a2f6deac823d0#diff-e126536bed52648bd8b876e5a90622a3ebfc00207b7c0b10ca7f1f9605e67c0eR114

martinzink commented on extensions/standard-processors/tests/unit/PutUDPTests.cpp:
```diff
@@ -24,91 +24,87 @@
+    REQUIRE(failure_flow_files.size() == 1);
+    CHECK(result.at(PutUDP::Success).empty());
+    CHECK(controller.plan->getContent(failure_flow_files[0]) == message);
+    CHECK((LogTestController::getInstance().contains("Host not found") || LogTestController::getInstance().contains("No such host is known")));
```
good idea, changed it in https://github.com/apache/nifi-minifi-cpp/pull/1412/commits/45948030831865d2a71d6a47135a2f6deac823d0#diff-f690439939269c49549e72231c05edf2e8ba912ec3fe206031d6465b7d161779R98

martinzink commented:
I've moved from only IPv4 only mode to IPv6, I also modified ListenTCP and ListenSyslog processors aswell.
This is effectively fixes a separate ticket aswell. So could you guys review the latest [commit](https://github.com/apache/nifi-minifi-cpp/pull/1412/commits/aba9c5bb58f317143a6301e1552087c8f6d4a5fa) aswell? @fgerlits @lordgamez

martinzink commented on extensions/standard-processors/processors/PutUDP.cpp:
```diff
@@ -107,48 +98,48 @@ void PutUDP::onTrigger(core::ProcessContext* context, core::ProcessSession* cons
+  const auto debug_log_resolved_endpoint = [&hostname, &logger = this->logger_](const udp::resolver::results_type& resolved_query) -> udp::endpoint {
+    if (logger->should_log(core::logging::LOG_LEVEL::debug))
+      core::logging::LOG_DEBUG(logger) << "resolved " << hostname << " to: " << resolved_query->endpoint();
+    return resolved_query->endpoint();
```
I've reworked this part (and also ListenTCP, ListenSyslog) in https://github.com/apache/nifi-minifi-cpp/pull/1412/commits/aba9c5bb58f317143a6301e1552087c8f6d4a5fa

martinzink commented on extensions/standard-processors/processors/PutUDP.cpp:
```diff
@@ -107,48 +98,48 @@ void PutUDP::onTrigger(core::ProcessContext* context, core::ProcessSession* cons
+  const auto send_data_to_endpoint = [&io_context, &data](const udp::endpoint& endpoint) -> nonstd::expected<void, std::error_code> {
+    std::error_code send_error;
+    udp::socket socket(io_context);
+    socket.open(udp::v4());
```
I've reworked this part (and also ListenTCP, ListenSyslog) in https://github.com/apache/nifi-minifi-cpp/pull/1412/commits/aba9c5bb58f317143a6301e1552087c8f6d4a5fa

fgerlits commented on libminifi/test/Utils.h:
```diff
@@ -111,24 +111,42 @@ bool countLogOccurrencesUntil(const std::string& pattern,
   for (auto& content : contents) {
     std::string tcp_message(content);
     tcp_message += '\n';
     asio::write(socket, asio::buffer(tcp_message, tcp_message.size()), err);
```
Won't this overwrite the previous value of `err`?  I.e., if there are 2 writes, the first fails and the second succeeds, then we'll return `true` as if everything was A-OK.

martinzink replied:
You are right :+1: I moved the error checking inside the loop in https://github.com/apache/nifi-minifi-cpp/pull/1412/commits/3caa5a47fc16dc44dddc298c7f3d20b33fefe6f7

martinzink commented on extensions/standard-processors/tests/unit/ListenSyslogTests.cpp:
```diff
@@ -197,21 +197,10 @@ constexpr std::string_view rfc5424_logger_example_1 = R"(<13>1 2022-03-17T10:10:
   CHECK(std::to_string(port) == flow_file.getAttribute("syslog.port"));
   CHECK(protocol == flow_file.getAttribute("syslog.protocol"));
-  CHECK("127.0.0.1" == flow_file.getAttribute("syslog.sender"));
+  CHECK(("::ffff:127.0.0.1" == flow_file.getAttribute("syslog.sender") || "::1" == flow_file.getAttribute("syslog.sender")));
```
the `::ffff:` is a subnet prefix for IPv4 addresses that are placed inside IPv6.
The UDP/TCP Servers now listen on IPv6 with dual stacking enabled, this means that the OS listens on ipv4 and ipv6 simultaneously and any request that comes through ipv4 will be transferred to ipv6's ipv4 subnet, thus `127.0.0.1` (original sender) will become `::ffff:127.0.0.1`

martinzink commented on libminifi/test/Utils.h:
```diff
@@ -111,24 +111,42 @@ bool countLogOccurrencesUntil(const std::string& pattern,
+  if (err)
+    return false;
+  socket.close();
+  return true;
```
Just checked and fortunetly RAII will take of everything so we dont need to manually close it anywhere.
I moved the error checking inside the loop in https://github.com/apache/nifi-minifi-cpp/pull/1412/commits/3caa5a47fc16dc44dddc298c7f3d20b33fefe6f7

martinzink commented on extensions/standard-processors/processors/PutUDP.cpp:
```diff
@@ -107,51 +98,53 @@ void PutUDP::onTrigger(core::ProcessContext* context, core::ProcessSession* cons
-          names_vector | ranges::views::join(',') | ranges::to<std::string>());
+      socket.send_to(asio::buffer(data.buffer), resolver_entry.endpoint(), udp::socket::message_flags{}, error);
+      if (error) {
+        core::logging::LOG_DEBUG(logger) << "sending to endpoint " << resolver_entry.endpoint() << " failed due to " << error.message();
```
As long as we succeed to send to one resolved endpoint, I don't think we need to handle these as warnings or errors, since this can happen during normal operation. (and the last error will be logged as error)

I've added the success logging in https://github.com/apache/nifi-minifi-cpp/pull/1412/commits/3caa5a47fc16dc44dddc298c7f3d20b33fefe6f7 and https://github.com/apache/nifi-minifi-cpp/pull/1412/commits/64dfce35ea2b477ecd8695bd868b184da9696512

szaszm commented on libminifi/test/Utils.h:
```diff
@@ -111,24 +111,47 @@ bool countLogOccurrencesUntil(const std::string& pattern,
   return true;
 }
 
+bool sendUdpDatagram(const asio::const_buffer& content, const asio::ip::udp::endpoint& remote_endpoint) {
```
`asio::const_buffer` is cheap to copy, similarly to `gsl::span` and `std::string_view`. I would pass it by value.

szaszm commented on libminifi/test/Utils.h:
```diff
@@ -141,7 +141,7 @@ bool sendUdpDatagram(const gsl::span<std::byte const> content, const asio::ip::u
 
 bool sendUdpDatagram(const std::string_view content, const asio::ip::udp::endpoint& remote_endpoint) {
-  return sendUdpDatagram(asio::const_buffer(content.begin(), content.size()), remote_endpoint);
+  return sendUdpDatagram(asio::buffer(content), remote_endpoint);
```
Is this const correct?

martinzink replied:
Yes it is, the naming of asio buffers is a bit confusing though...
there are two classes, [asio::const_buffer](https://github.com/chriskohlhoff/asio/blame/master/asio/include/asio/buffer.hpp#L227), and [asio::mutable_buffer](https://github.com/chriskohlhoff/asio/blame/master/asio/include/asio/buffer.hpp#L93).
asio::buffer is only a function, in this case asio::buffer(std::string_view) creates an asio::const_buffer https://github.com/chriskohlhoff/asio/blame/master/asio/include/asio/buffer.hpp#L1516-L1528
